src/services/services.ts

```diff
@@ -6259,6 +6379,193 @@ namespace ts {
+                }
+
+                return node.kind === SyntaxKind.ArrowFunction ||
+                    node.kind === SyntaxKind.FunctionExpression ||
+                    node.kind === SyntaxKind.ObjectLiteralExpression ||
+                    node.kind === SyntaxKind.ClassExpression;
+            }
+
+            /**
+             * Determines if the parent symbol occurs somewhere in the child's ancestry. If the parent symbol
+             * is an interface, determines if some ancestor of the child symbol extends or inherits from it.
+             * This also takes in a cache of previous results which makes this slightly more efficient and is
+             * necessary to avoid potential loops like so:
+             *     class A extends B { }
+             *     class B extends A { }
+             *
+             * @param child         A class or interface Symbol
+             * @param parent        Another class or interface Symbol
+             * @param cachedResults A map of symbol names to booleans indicating previous results
+             */
+            function inheritsFrom(child: Symbol, parent: Symbol, cachedResults: Map<boolean> = createMap<boolean>()): boolean {
+                const parentIsInterface = parent.getFlags() & SymbolFlags.Interface;
+                return searchHierarchy(child, cachedResults);
+
+                function searchHierarchy(symbol: Symbol, cachedResults: Map<boolean>): boolean {
+                    if (symbol === parent) {
+                        return true;
+                    }
+                    else if (symbol.name in cachedResults) {
+                        return cachedResults[symbol.name];
+                    }
+
+                    cachedResults[symbol.name] = false;
+
+                    const inherits = forEach(symbol.getDeclarations(), (declaration) => {
+                        if (isClassLike(declaration)) {
+                            if (parentIsInterface) {
+                                const interfaceReferences = getClassImplementsHeritageClauseElements(declaration);
+                                if (interfaceReferences) {
+                                    for (const typeReference of interfaceReferences) {
+                                        if (searchTypeReference(typeReference, cachedResults)) {
+                                            return true;
+                                        }
+                                    }
+                                }
+                            }
+                            return searchTypeReference(getClassExtendsHeritageClauseElement(declaration), cachedResults);
+                        }
+                        else if (declaration.kind === SyntaxKind.InterfaceDeclaration) {
+                            if (parentIsInterface) {
+                                return forEach(getInterfaceBaseTypeNodes(<InterfaceDeclaration>declaration), base => searchTypeReference(base, cachedResults));

```

Comment:
cacheResults is already captured in the closure, so you do not need to pass it along. consider making searchTypeReference take one argument, this way you do not need the lambda.